### DISCUSSION TITLE
Preprocess outputs, resolutions and resampling

### DISCUSSION THREAD

**Discussion by biancadepifanio** [CoBrALab/RABIES#275] [Q&A] [answered]
Hi guys :) First of all, thank you for having created such a tool!
I have some questions about the registration steps and it would be amazing if you can help me in understanding better and clarify the output of the core preprocessing.
I run this command here:  
```
docker run -it --rm \
-v $PWD/input_BIDS:/input_BIDS:ro \
-v $PWD/preprocess_outputs:/preprocess_outputs/ \
gabdesgreg/rabies:0.4.8 -p MultiProc --local_threads 6 --scale_min_memory 5 preprocess \
/input_BIDS/ /preprocess_outputs/ \
--bold_only \
--apply_despiking  \
--detect_dummy \
--TR 1 \
--bold_inho_cor method=SyN,otsu_thresh=2,multiotsu=false \
--commonspace_reg masking=true,brain_extraction=false,template_registration=SyN,fast_commonspace=false


```
My **input image** had this resolution here: [input_hd.txt](https://github.com/CoBrALab/RABIES/files/10841488/input_hd.txt)

From my understanding, the first step of your robust registration workflow is the creation of an **initial bold reference image** (i.e. initial_bold_ref ), whose resolution is the same as the input image ([initial_bold_ref_hd.txt](https://github.com/CoBrALab/RABIES/files/10841497/initial_bold_ref_hd.txt))
After that, RABIES creates the **inho_bold_cor** image that has this header: [inho_cor_bold_hd.txt](https://github.com/CoBrALab/RABIES/files/10841504/inho_cor_bold_hd.txt), and it is produced after inhomogeneity correction and later used for registration of the EPI.
At this point from the single-scan space, RABIES passes to the registration in the common space. In my outputs, there is also the **resampled_template** (with this header [commonspace_resampled_template_hd.txt](https://github.com/CoBrALab/RABIES/files/10841513/commonspace_resampled_template_hd.txt)), which I think is created after resampling the **EPI_template** found here [RABIES_extras | Zenodo](https://zenodo.org/record/5118030#.Y_eQJibMKUk) (with this header:[EPI_template_hd.txt](https://github.com/CoBrALab/RABIES/files/10841516/EPI_template_hd.txt)) to the input EPI image.  
Finally, I have the **EPI_bold_combined** image with this header: [commonspace_EPI_bold_combined_hd.txt](https://github.com/CoBrALab/RABIES/files/10841519/commonspace_EPI_bold_combined_hd.txt) – same voxel dimension as the input image, but cropped FOV that doesn’t match with any of the intermediate images nor with the EPI_template. Moreover, I’d have expected to have the “unbiased” template available with the same resolution and FOV of the final EPI_BOLD_combined.
In conclusion, at the end of the core preprocessing (-- bold_only) I have six different images (Input image, initial_bold_ref, inho_cor_bold, EPI_template, resampled_template and EPI_bold_combined), each one with a different FOV (resulting, thus, in different coordinate-spaces).

Can you please clarify a bit better what each output represents and how it was obtained?

Thank you!

**Comment by Gab-D-G** [comment]
Hi, the documentation does not currently cover the steps at this level of details, so these are all good questions to clarify. 

- As you mentioned, the **input image** and **initial_bold_ref** will always have the same, as neither gets resampled
- The input template image defining the commonspace gets resampled either according to user-specified dimensions with --anatomical_resampling, or by default, it gets matched to the lowest dimension of the set of input images for commonspace alignments (i.e. the set of structural scans if provided, or the set of EPIs if using the --bold_only option), to avoid increasing image resolution to match the external template during registration, which would increase computational time. The resampling function is here https://github.com/CoBrALab/RABIES/blob/6f19f204c984ed135c101936e60422ed2e06651a/rabies/preprocess_pkg/utils.py#L182. So, the **resampled_template** is the template file actually used internally by the pipeline, and the resolution will likely differ from the **EPI_template**. Importantly, the resampled template defines the resolution at which registration operations are conducted across the pipeline, and thus computational time is proportional to that resolution. 
- During inhomogeneity correction, either BOLD or anatomical, the input image gets first resampled to match the resolution of the template in commonspace; thus the **inho_cor_bold** will match the resolution of **resampled_template**. Furthermore, when a mask is inherited during inhomogeneity correction, the FOV of the image gets cropped to the surrounding of the brain mask, so the FOV of **inho_cor_bold** will be smaller than the **input image**.
- Finally, after registration to commonspace is completed, the original EPI image gets resampled onto that commonspace, but while preserving the original resolution, as we don't want to change the resolution of the actual timeseries that get analyzed later (unless otherwise specified by --commonspace_resampling or --nativespace_resampling). However, the FOV of the resampled image will match the target image on which it gets resampled (the commonspace template FOV for commonspace outputs, and the anatomical scan for native space outputs). Thus, **EPI_bold_combined** will have the resolution of the original image, but the FOV of the commonspace template in your case.

Hope this helps! Let me know if you have more questions.
Gabriel

**Comment by biancadepifanio** [comment]
Hi Gabriel, thank you so much for your answer, it was very helpful :) 

I still have some doubts about the template in the commonspace. I would have expected to have the commonspace template in the same space as the final preprocessed image, instead the resampled_template is in another space with respect to the final preprocessed images. (different FOV). I think I am missing how from the inho_cor_bold image that was resampled on the resampled_template, we arrive to the EPI_bold_combined image, that has the same resolution as the input and different from the inho_cor_bold and resampled_template. Moreover, the EPi_bold_combined image (in my case) doesn't have same FOV of the commonspace image. It is like I am missing one intermediate step and the "real" commonspace template.
I hope I expressed clearly my doubt, 

Thank you for your help! 
Bianca :)

**Comment by Gab-D-G** [comment, marked as answer]
Hi Bianca, 
by commonspace template you are referring to the file found in bold_datasink/commonspace_resampled_template/, and comparing it to the bold_datasink/commonspace_bold preprocessed EPI images? These two files should have the same FOV, but can have different resolutions.

As for your question about resampling, the final step takes the transforms previously computed from the registration steps (from the inho_cor_bold, head motion realignment, etc.), and applies them to your input timeseries with `antsApplyTransforms`, for each one EPI frame at a time. With this step, we also control at which resolution we want the EPI to be resampled to, and by default, RABIES constrains the resampling to retain the original voxel resolution. At the same time, a target image must be provided during resampling to define the FOV in that new space, which will be the commonspace image for the commonspace_bold outputs. Hope this helps!

Best,
Gabriel DG

**Comment by biancadepifanio** [comment]
Hi Gabriel, 

Thank you. All it's clear now :)

**Comment by Gab-D-G** [comment]
Hi Bianca, I made a small mistake in my original comment, where I specified that the template was resampled to match the lowest resolution of the EPI image, but this is actually only if using the --bold_only option (i.e. not using structural scans). If using structural images, the template is resampled to match the lowest resolution of the structural images instead. I corrected the original comment. - Gabe
